### PR TITLE
Implement tasks feature

### PR DESCRIPTION
**Pull request by nicksellen**
It works!

Features:
- a task list per _stage_ and per _topic_
- can add them
- can assign to any river swimmer if a _starter_, otherwise just yourself
- due date is optional
- can mark as done :)
- can edit
- can delete
- tasks is a foldable section
- ordered by due date, then created date

[Kooha-2023-09-15-17-46-47.webm](https://github.com/animorphcoop/shared-futures-space/assets/31616/261208ba-fc76-4769-b041-83419439986e)

Dev/tech stuff:
- I added two new alpine directives:
  - `x-stop-body-scroll` (magic!)
  - `x-characters-remaining`
- I created HTMXMixin so views can work with full page loading _or_ htmx

TODO:
- [x] check permissions for starters vs normal users
- [ ] add wise salmon tips for useful tasks
- [x] decide if the native datepicker is good enough
- [x] squash the migrations to make one clean one
- [x] check there is no outstanding validation issues
- [ ] ~~add a loading indicator when saving?~~
- [ ] ~~remove full page views (which are just there for easier dev experience, maybe they can go somewhere dev-only?)~~
- [x] trawl through TODOs I left in the code
- [x] add tests (can we actually test full page interactions?)

Observations on the side:
- `package-lock.json` should be checked into repo
  - remove the ignore and add it in
- `LoginRequiredMixin` is useful :)
- I got my head into the htmx-mindset a lot now, other than the directives, I didn't write any js/ts 🥳

### CODE REVIEW

File: apps/core/templatetags/custom_filters.py [hunk 1]
```diff
@@ -18,3 +21,14 @@ def attrmap(value: Iterable[T], arg: str) -> List[T]:
 @register.filter(name="strcat")
 def strcat(value, arg):
     return str(value) + str(arg)
+
+
+@register.filter(name="to_json")
+@mark_safe
+def to_json(value):
+    return json.dumps(value)
```
Comment (nicksellen) [inline apps/core/templatetags/custom_filters.py]: This is unused now, can remove

File: apps/task/migrations/0005_task_created_at.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        ('task', '0004_alter_task_due'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='task',
+            name='created_at',
+            field=models.DateTimeField(default=django.utils.timezone.now),
+        ),
+    ]
```
Comment (nicksellen) [inline apps/task/migrations/0005_task_created_at.py]: Squash all these migrations into one neat one...

File: apps/task/templatetags/task_tags.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    if slug and stage and topic:
+        return "-".join(
+            [
+                "task-list",
+                slug,
+                stage,
+                topic,
+            ]
+        )
+    return None
```
Comment (nicksellen) [inline apps/task/templatetags/task_tags.py]: This has potential to be a more general utility for creating ids that make use of slug/stage/topic....

File: apps/task/views.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    def form_valid(self, form) -> HttpResponse:
+        form.instance.river = self.get_river()
+        form.instance.stage_name = self.get_stage_name()
+        form.instance.topic = self.get_topic()
+        return super().form_valid(form)
+
+    def form_invalid(self, form) -> HttpResponse:
+        # TODO: there might be things we don't handle here very well...
+        print("FORM INVALID")
```
Comment (nicksellen) [inline apps/task/views.py]: Check this!

File: apps/task/views.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        return redirect(self.get_success_url())
+
+
+class DeleteTaskView(LoginRequiredMixin, TaskViewMixin, DeleteView):
+    model = Task
+
+    template_name_htmx = "task/partials/task_delete.html"
+
+    def get_object(self):
+        return self.get_queryset().get(uuid=self.kwargs["uuid"])
```
Comment (nicksellen) [inline apps/task/views.py]: Nice small view classes using the django base ones, `DeleteView`, etc...

File: requirements_dev.txt [hunk 6]
```diff
@@ -1,3 +1,3 @@
-black==23.3.0
+black[d]==23.3.0
```
Comment (nicksellen) [inline requirements_dev.txt:1]: `[d]` variant is to support black daemon, which the `BlackConnect` intellij plugin I use uses...

File: templates/messaging/chatbox_snippet.html [hunk 7]
```diff
@@ -93,8 +93,10 @@
                     if (!uploadCheckbox.classList.contains('hidden')) {
                         uploadCheckbox.classList.add('hidden')
                     }
-                    const elem = document.getElementById(formId).closest('.drawer-contents')
+                    // We want to scroll the messages list to the bottom
+                    const elem = document.getElementById(targetId)
                     if (elem) {
+                        window.elem = elem
```
Comment (nicksellen) [inline templates/messaging/chatbox_snippet.html]: Remove this, was just for debugging something

**Comment by nicksellen** [comment]
I did the things, still no wise salmon tips, as I'm not sure what the content is supposed to be, I think they are a bit underspecified, or I missed the bit where they are specified :), either way I think it makes more sense to merge this first.

File: templates/ts/directives/characters-remaining.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * A directive to print the characters remaining, usage guide:
+ *
+ *    <input id="foo" type="text" maxlength="200">
+ *    <span x-characters-remaining="#foo"></span> characters remaining
+ *
+ * This will show: "200/200 characters remaining" and update as you type.
+ */
+Alpine.directive('characters-remaining', (
```
Comment (sirodoht) [inline templates/ts/directives/characters-remaining.ts:11]: that's cool!

**Review by sirodoht — APPROVED** [review]
Looks great! Sorry, I rushed the other PR merge and this one has some conflicts now but I think it won't be too bad because I simplied the settings!

**Comment by nicksellen** [comment]
Ok, rebased ontop of development again :)

**Review by sirodoht — APPROVED** [review]
👌

**Comment by sirodoht** [comment]
🎉

**Pull request merged**